pkg/controller/elasticsearch/filesettings/secret.go

```diff
@@ -0,0 +1,191 @@
+// belongs to no one or because it already belongs to the given policy.
+func (s SettingsSecret) CanBeOwnedBy(policy policyv1alpha1.StackConfigPolicy) (reconciler.SoftOwnerRef, bool) {
+	currentOwner, referenced := reconciler.SoftOwnerRefFromLabels(s.Labels)
+	// either there is no soft owner
+	if !referenced {
+		return reconciler.SoftOwnerRef{}, true
+	}
+	// or the owner is already the given policy
+	canBeOwned := currentOwner.Kind == policy.Kind && currentOwner.Namespace == policy.Namespace && currentOwner.Name == policy.Name
+	return currentOwner, canBeOwned
+}
+
+// setSoftOwner sets the given StackConfigPolicy as soft owner of the Settings Secret using the "softOwned" labels.
+func (s *SettingsSecret) setSoftOwner(policy policyv1alpha1.StackConfigPolicy) {
+	if s.Labels == nil {
+		s.Labels = map[string]string{}
+	}
+	s.Labels[reconciler.SoftOwnerNamespaceLabel] = policy.GetNamespace()
+	s.Labels[reconciler.SoftOwnerNameLabel] = policy.GetName()
+	s.Labels[reconciler.SoftOwnerKindLabel] = policy.GetObjectKind().GroupVersionKind().Kind
+}
+
+// setSecureSettings stores the SecureSettings Secret sources referenced in the given StackConfigPolicy in the annotation of the Settings Secret.
+func (s *SettingsSecret) setSecureSettings(policy policyv1alpha1.StackConfigPolicy) error {
+	if len(policy.Spec.SecureSettings) == 0 {
+		return nil
+	}
+
+	secretSources := make([]commonv1.NamespacedSecretSource, len(policy.Spec.SecureSettings))
+	for i, src := range policy.Spec.SecureSettings {
+		secretSources[i] = commonv1.NamespacedSecretSource{Namespace: policy.GetNamespace(), SecretName: src.SecretName, Entries: src.Entries}
+	}
+
+	bytes, err := json.Marshal(secretSources)
+	if err != nil {
+		return err
+	}
+	s.Annotations[secureSettingsSecretsAnnotationName] = string(bytes)
+	return nil
+}
+
+// getSecureSettings returns the SecureSettings Secret sources stores in an annotation of the Secret.
+func (s *SettingsSecret) getSecureSettings() ([]commonv1.NamespacedSecretSource, error) {
+	rawString, ok := s.Annotations[secureSettingsSecretsAnnotationName]
+	if !ok {
+		return []commonv1.NamespacedSecretSource{}, nil
+	}
+	var secretSources []commonv1.NamespacedSecretSource
+	err := json.Unmarshal([]byte(rawString), &secretSources)
+	if err != nil {
+		return nil, err
+	}
+	return secretSources, nil
+}
+
+// GetSecureSettingsSecretSources gets SecureSettings Secret sources for a given Elastic resource.
+func GetSecureSettingsSecretSources(ctx context.Context, c k8s.Client, resource metav1.Object) ([]commonv1.NamespacedSecretSource, error) {
+	var secret corev1.Secret
+	err := c.Get(ctx, types.NamespacedName{Namespace: resource.GetNamespace(), Name: esv1.FileSettingsSecretName(resource.GetName())}, &secret)
+	if apierrors.IsNotFound(err) {
+		return []commonv1.NamespacedSecretSource{}, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	settingsSecret, err := NewSettingsSecretFromSecret(secret)
```

Comment:
We unmarshal the whole secret here only but we only need the secure settings which are stored in an annotation? If we removed the receiver from `getSecureSettings()` we could skip that first step.